libazureinit/src/config.rs

```diff
@@ -0,0 +1,820 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! Module for configuring azure-init.
+//!
+//! This module provides configuration structures and methods for loading and merging
+//! configurations from files or directories. Configurations can be customized using
+//! `Config` struct options to define settings for SSH, hostname provisioners, user
+//! provisioners, IMDS, provisioning media, and telemetry.
+use crate::error::Error;
+use figment::{
+    providers::{Format, Serialized, Toml},
+    Figment,
+};
+use serde::{Deserialize, Serialize};
+use std::path::PathBuf;
+use std::{fmt, fs};
+use toml;
+use tracing;
+
+#[derive(Default, Serialize, Deserialize, Debug, Clone, Copy, PartialEq)]
+#[serde(rename_all = "lowercase")]
+#[non_exhaustive]
+pub enum HostnameProvisioner {
+    #[default]
+    Hostnamectl,
+    #[cfg(test)]
+    FakeHostnamectl,
+}
+
+#[derive(Default, Serialize, Deserialize, Debug, Clone, Copy, PartialEq)]
+#[serde(rename_all = "lowercase")]
+pub enum UserProvisioner {
```

Comment:
Sorry, I should have explicitly called this out, but we'll want the same fix-up you did for https://github.com/Azure/azure-init/pull/137/files#r1854279256 to this enum as well as the PasswordProvisioner enum below.

Reply:
No, that's on me! Should have caught that, thank you! Will fix them all:)